src/rust-crypto/RoomEncryptor.ts

```diff
@@ -91,7 +103,33 @@ export class RoomEncryptor {
             );
         }
 
+        // Manually call `loadMembersIfNeeded` here, because we want to know if it's the first
+        // time the room is loaded (due to lazy loading), so we can update the tracked users.
+        const fromServer = await this.room.loadMembersIfNeeded();
         const members = await this.room.getEncryptionTargetMembers();
+
+        if (fromServer) {
+            // It's the first time the room is loaded, so we need to update the tracked users
+            await this.olmMachine
+                .updateTrackedUsers(members.map((u) => new RustSdkCryptoJs.UserId(u.userId)))
+                .then(() => {
+                    this.prefixedLogger.debug(`Updated tracked users for room ${this.room.roomId}`);
+                });
+        }
+
+        // Query keys in case we don't have them for newly tracked members.
+        // This must be done before ensuring sessions. If not the devices of these users are not
+        // known yet and will not get the room key.
+        // We don't have API to only get the queries related to this member list, so we just
+        // process the pending `KeysQuery` requests from the olmMachine. (usually these are processed
+        // at the end of the sync, but we can't wait for that).
+        const request: OutgoingRequest[] = (await this.olmMachine.outgoingRequests()).filter(
+            (r: OutgoingRequest) => r.type === RequestType.KeysQuery,
+        );
+        for (let i = 0; i < request.length; i++) {
+            await this.outgoingRequestProcessor.makeOutgoingRequest(request[i]);
+        }
```

Comment:
danger here is that we could end up making the `/keys/query` requests lots of times concurrently (once in the `outgoingRequestLoop`, once for each call to `ensureEncryptionSession`).

It might be best to do some refactoring of `outgoingRequestLoop` to allow us to wait for a single iteration of the loop (and not worry too much about limiting it to `KeysQuery` requests).

Reply:
I have extracted the loop logic in OutgoingRequestsManager and added tests